Rust/src/algorithm/visitor.rs

```diff
@@ -0,0 +1,38 @@
+use crate::{Internal, Leaf};
+
+/// Visitors on random cut trees.
+///
+/// Many random cut forest scoring algorithms use a visitor design pattern on
+/// the nodes of a tree. The nodes typically run from the leaf closest to the
+/// input query point along the path to the root node such as in the function
+/// [`Tree::traverse()`](crate::tree::Tree::traverse).
+///
+pub trait Visitor<T> {
+
+    /// The output type of the visitor.
+    type Output;
+
+    /// Initialize visitor calculation at the leaf node nearest a query point.
+    ///
+    /// This function is called on the leaf node nearest to an input query
```

Comment:
Because the leaf is passed in, the description shouldn't allude to which leaf is passed in. I do not think it necessary here to make assumptions on that leaf.

Reply:
Good call. I agree that the visitor is agnostic to the traversal algorithm.